### PR TITLE
Plane: Quadplane: add new mode: loiter to alt then Qland.

### PR DESCRIPTION
**Pull request by IamPete1**
This work is supported by ARACE.

This adds a new mode that will loiter to QRTL altitude in forward flight then switch to QLAND. For use when a QLAND would result in trying to loiter down from a large altitude. If in VTOL flight it will switch straight to QLAND, if in forward flight it will loiter down first. 

This is just the mode, still to be done is adding it to the various fail-safe options, mode switches ect.

### CODE REVIEW

**Comment by tridge** [comment]
thanks!
I wondered about making this the default for QLAND based on a configurable alt threshold. Default threshold could be zero, meaning current behaviour.

**Comment by tridge** [comment]
it would also be good to be able to move the landing point on GCS, like the fly-to for guided, so if its heading for a bad spot and you have time you can easily say "head for that clearing over there"

**Comment by pompecukor** [comment]
> 
> 
> thanks!
> I wondered about making this the default for QLAND based on a configurable alt threshold. Default threshold could be zero, meaning current behaviour.

Some times you want Qland as qland, even from 100m. I can give some examples.
However other times Loiter2Qland is more desirable.
Mostly both are for emergencies and in emergencies typically you do not have time for changing parameters.

Examples.
Typically I would expect if this "new mode" went in, then everyone (in their right minds), especially with hybrid battery setup (primary for pusher and secondary for Quad). And also for ICE like Alti and such. Would want to send BAT_CRT_ACTION to this Loit_to_Qland

Then for example in a testing or where we have know location as clear, but something wrong with forward flight system (like failed elevator servo) or something else is with forward flight it amis, even if it is in forward flight and we are too far to QRTL, then QLAND is desirable. I mean you can QLAND from 300m if you have enough battery, we have done it.

However in both examples above it is not really optimal to have to go into parameters to change something, when in fact all we want it to select the desired action by switching a mode on Mission Planner or configured RC switch.
Yes we can use scripting for more "intelligent" weighing, but our primary desire for this is to have as an individual mode.
It just makes sense.

**Comment by pompecukor** [comment]
> 
> 
> it would also be good to be able to move the landing point on GCS, like the fly-to for guided, so if its heading for a bad spot and you have time you can easily say "head for that clearing over there"

Yes, you did mention this when we first discussed the idea of this new mod. Obviously I do not mind at all, however I think don't know how you would set such up. So I am not against, but hope that it will not delay it. If you find a sensible way to add that.
However please note that right now for us to exactly what you describe it very easy with current firmware.
You know we use Q_RLT_MODE: 2.
So as in your example, if I am in an emergency, but still have time and resources (map, camera, etc) to know where I want to land. Then it is so easy to just right click on that point in Mission Planner map, then select "set home here" (see picture below).
After that all I need to do is hit RTL. The craft will perform the desired action, fly/loiter to alt over the point, then at Q_RTL_ALT if will transition and land on the spot where we set the new "home".
Is this not what you are after? This works now.

image

**Comment by peterbarker** [comment]
Couldn't we use guided mode for this?

`MAV_CMD_NAV_VTOL_LAND` appears to already have the desired behaviour; it takes a lat/lng and will fly fixed-wing to that location if required.

This is roughly similar to MAV_CMD_DO_REPOSITION`, which forces guided mode if you're not already in it.

As opposed to the above, however, GCS changes would be required - instead of "right-click, Set Home Here" and then RTL, it would be "right-click, select "Land Here", which would issue `MAV_CMD_NAV_VTOL_LAND` to the autopilot.

File: libraries/AP_Common/Location.cpp [hunk 1]
```diff
@@ -244,6 +244,16 @@ ftype Location::get_distance(const struct Location &loc2) const
     return norm(dlat, dlng) * LOCATION_SCALING_FACTOR;
 }
 
+// return the altitude difference in meters taking into account alt frame.
```
Comment (peterbarker) [inline libraries/AP_Common/Location.cpp:247]: We should be clear about which way around this is - e.g. `distance will be positive if loc2 is above *this` (I'm not even sure that's correct :-) )

Comment (IamPete1) [reply libraries/AP_Common/Location.cpp:247]: I copied the why we do it in distance NED, but as you say there both confusing.

**Comment by pompecukor** [comment]
> 
> 
> Couldn't we use guided mode for this?
> 
> `MAV_CMD_NAV_VTOL_LAND` appears to already have the desired behaviour; it takes a lat/lng and will fly fixed-wing to that location if required.
> 
> This is roughly similar to MAV_CMD_DO_REPOSITION`, which forces guided mode if you're not already in it.
> 
> As opposed to the above, however, GCS changes would be required - instead of "right-click, Set Home Here" and then RTL, it would be "right-click, select "Land Here", which would issue `MAV_CMD_NAV_VTOL_LAND` to the autopilot.

I am sure there are many ways to accomplish this one action that Tridge mentioned. My point was that this works currently and works well. Without need to change anything in MP either. So not sure there is a point is changing it all around.

**Comment by IamPete1** [comment]
Rebased, added as a battery failsafe option and added support for "fly-to" from GCS

File: ArduPlane/Plane.h [hunk 2]
```diff
@@ -299,6 +300,9 @@ class Plane : public AP_Vehicle {
 #if HAL_SOARING_ENABLED
     ModeThermal mode_thermal;
 #endif
+#if HAL_QUADPLANE_ENABLED
+    ModeLoiterAltQLand mode_lotier_qland;
```
Comment (peterbarker) [inline ArduPlane/Plane.h:304]:
```suggestion
    ModeLoiterAltQLand mode_loiter_qland;
```

**Pull request merged**